extensions/script-libraries/mime/mime.livecodescript

```diff
@@ -0,0 +1,1271 @@
+      pass revUnloadLibrary
+   end if
+   
+   remove the script of me from back
+end revUnloadLibrary
+
+/**
+
+This script library implements portions of Multipurpose Internet Mail Extensions (MIME)
+as described in RFCs 2045 - 2049
+
+*/
+
+local sMIMETypes
+
+/**
+
+The MIME version this implemented in the library
+
+Example:
+put "MIME-Version:" && MIMEVersion() & crlf after tHeaders
+
+Description:
+Use the <MIMEVersion> function create a `MIME-Version` header.
+
+*/
+
+function MIMEVersion
+   return 1.0
+end MIMEVersion
+
+/**
+
+Determine the MIME type for a file extension
+
+Example:
+put "Content-Type:" && MIMETypeForExtension("xls") & crlf after tHeaders
+
+Parameters:
+pExtension (string):
+A file extension
+
+Description:
+The mime types listed are the default mime types in an Apache web server installation
+for a given file extension. If the file extension is not found then `application/octet-stream`
+will be returned.
+
+Returns:
+A mime type for the given file extension.
+
+*/
+
+function MIMETypeForExtension pExtension
+   local tMIMEType
+   put sMIMETypes[pExtension] into tMIMEType
+   if tMIMEType is empty then
+      return "application/octet-stream" for value
+   else
+      return tMIMEType for value
+   end if
+end MIMETypeForExtension
+
+/**
+
+List extensins that match a MIME type or MIME type wildcard pattern
+
+Example:
+put line 1 of MIMETypeExtensions("application/epub+zip") into tExtension
+
+Example:
+put MIMETypeExtensions("image/*") into tImageExtensions
+
+Parameters:
+pMIMEType (string):
+A MIME type or wildcard pattern for matching a MIME type
+
+Description:
+List the extensions associated with a given MIME type or with a MIME type matching
+a wildcard pattern.
+
+Returns:
+A list of extensions or empty if none were found.
+
+*/
+
+function MIMETypeExtensions pMimeType
+   local tExtensions
+   filter elements of sMIMETypes with wildcard pattern pMimeType into tExtensions
+   
+   local tExtensionList
+   repeat for each key tExtension in tExtensions
+      put true into tExtensionList[tExtension]
+   end repeat
+   
+   combine tExtensionList with return as set
+   sort tExtensionList
+   
+   return tExtensionList
+end MIMETypeExtensions
+
+/**
+
+Encode data in content transfer encoding
+
+Example:
+put MIMEEncode("quoted-printable", tEmailBody, "UTF-8") after tData
+
+Parameters:
+pEncoding (enum):
+The encoding to return the data in
+- "quoted-printable": Quoted printable encoding. Most likely used for text with many ASCII characters
+- "base64": Base 64 encoding
+
+pData:
+The data to encode. Treated as binary data if the pTextEncoding parameter is empty
+otherwised treated as text to be encoded with the specified encoding.
+
+pTextEncoding: An encoding supported by the <textEncode> function
+
+Description:
+Use <MIMEEncode> to encode data for transfer
+
+Returns:
+The encoded content
+
+References:
+textEncode (function)
+
+*/
+
+function MIMEEncode pEncoding, pData, pTextEncoding
```

Comment:
Consider re-ordering the arguments to this function.  It feels a bit weird to have the two encodings on either side of the input data parameter.

What about either making the `pTextEncoding` parameter optional or even removing it altogether?  We can now automatically detect whether the argument is binary data or a string by using the `is strictly a binary string` predicate, and nowadays it seems reasonable to default to encoding strings as UTF-8 for e-mail.  If the prototype was:

```
function MIMEEncode pData, pEncoding
```

then it doesn't seem to awkward to write:

```
-- The defaults are fine
MIMEEncode("my text", "quoted-printable")
-- For some reason my heart's desire is UTF-16-BE
MIMEEncode(textEncode("my text", "UTF-16-BE"), "quoted-printable")
```